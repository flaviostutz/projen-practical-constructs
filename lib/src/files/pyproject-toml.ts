/* eslint-disable no-restricted-syntax */
import { stringify } from '@iarna/toml';
import { Dependency, DependencyType, FileBase, IResolver, Project } from 'projen';

type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] };

/** Package name format as in https://packaging.python.org/en/latest/specifications/name-normalization/ */
const PACKAGE_NAME_REGEX = /^([a-z0-9]|[A-z0-9][a-z0-9._-]*[a-z0-9])$/;

const VERSION_REGEX = /^(\d+\.\d+\.\d+)(-.+)?$/;

export type OptWithRequired = WithRequired<PyProjectTomlOptions, 'packageName' | 'version'>;

/**
 * pyproject.toml synthetisation
 */
export class PyProjectTomlFile extends FileBase {
  private readonly opts: OptWithRequired;

  constructor(project: Project, opts?: PyProjectTomlOptions) {
    super(project, 'pyproject.toml');
    this.opts = getOptionsWithDefaults(project, opts);
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  protected synthesizeContent(_resolver: IResolver): string | undefined {
    // get dependencies from project scope
    const dependencies = [];
    const devDependencies = [];
    for (const dep of this.project.deps.all) {
      if (dep.type === DependencyType.RUNTIME) {
        dependencies.push(renderDependency(dep));
      } else if (dep.type === DependencyType.DEVENV || dep.type === DependencyType.TEST) {
        devDependencies.push(renderDependency(dep));
      }
    }

    const contents = {
      'build-system': {
        requires: ['setuptools'],
        'build-backend': 'setuptools.build_meta',
      },
      project: {
        name: this.opts.packageName ?? '',
        version: this.opts.version ?? '',
        description: this.opts.description ?? '',
        readme: this.opts.readme ?? '',
        'requires-python': this.opts.requiresPython ?? '',
        license: '<LICENSE>',
        keywords: this.opts.keywords ?? [],
        // maintainers: this.opts.maintainers ?? [],
        dependencies: dependencies ?? [],
      },
      // 'project.optional-dependencies'
      projectOptionalDependencies: {
        dev: devDependencies ?? [],
      },
      // 'tool.setuptools.package-data'
      toolSetuptoolsPackageData: {
        packageDataPackageName: ['py.typed'],
      },
    };

    // we use those replacements becuse the toml library doesn't support these characters or nesting structures
    const tomlContents = stringify(contents)
      .replace('"<LICENSE>"', `{ file = "LICENSE" }`)
      .replace('projectOptionalDependencies', 'project.optional-dependencies')
      .replace('toolSetuptoolsPackageData', 'tool.setuptools.package-data')
      .replace('packageDataPackageName', `${this.opts.packageName}`);

    return tomlContents;
  }
}

const getOptionsWithDefaults = (project: Project, opts?: PyProjectTomlOptions): OptWithRequired => {
  if (opts?.version && !VERSION_REGEX.test(opts?.version)) {
    throw new Error(
      `Invalid version format (${opts?.version}). Must be in the format "X.Y.Z-<optional suffix>"`,
    );
  }

  return {
    requiresPython: '>=3.12',
    packageName: resolvePackageName(project.name, opts),
    version: '0.0.1',
    ...opts,
  };
};

export const resolvePackageName = (
  projectName: string,
  pyProjectOpts?: PyProjectTomlOptions,
): string => {
  let packageName = pyProjectOpts?.packageName;
  if (!packageName) {
    packageName = projectName;
  }
  if (!PACKAGE_NAME_REGEX.test(packageName!)) {
    throw new Error(
      `Invalid package name "${packageName}". Check requirements at https://packaging.python.org/en/latest/specifications/name-normalization/`,
    );
  }
  return packageName;
};

const renderDependency = (dep: Dependency): string => {
  let { version } = dep;
  // version starts with number, so add "==" because it has no qualifier
  if (version && /^[0-9]+.*/.test(version)) {
    version = `==${version}`;
  }
  return `${dep.name}${version ?? ''}`;
};

export interface PyProjectTomlOptions {
  /**
   * Name of the python package. E.g. "my_python_package".
   * Must only consist of alphanumeric characters and underscores.
   * @default Name of the directory
   */
  readonly packageName?: string;
  /**
   * Author's name
   */
  readonly authorName?: string;
  /**
   * Author's e-mail
   */
  readonly authorEmail?: string;
  /**
   * Version of the package.
   */
  readonly version?: string;
  /**
   * Keywords to add to the package.
   */
  readonly keywords?: string[];
  /**
   * List of maintainers of the package.
   */
  // readonly maintainers?: { name: string; email: string }[];
  /**
   * A short description of the package.
   */
  readonly description?: string;
  /**
   * A URL to the website of the project.
   */
  readonly homepage?: string;
  /**
   * README file.
   */
  readonly readme?: string;
  /**
   * License file
   */
  readonly licenseFile?: string;
  /**
   * Python version required to run this package
   */
  readonly requiresPython?: string;
  /**
   * A list of PyPI trove classifiers that describe the project.
   *
   * @see https://pypi.org/classifiers/
   */
  readonly classifiers?: string[];
}
