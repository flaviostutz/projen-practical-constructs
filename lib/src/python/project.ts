/* eslint-disable no-restricted-syntax */
/* eslint-disable no-new */

import { DependencyType, Project, ProjectOptions } from 'projen';
import { Projenrc } from 'projen/lib/python';

import { NODE_VERSION, PROJEN_VERSION } from '../common/constants';
import { BaseTooling } from '../common/components/base-tooling';

import { resolvePackageName } from './files/pyproject-toml';
import { PythonBasicSample } from './components/sample';
import { Build0Options, BuildTarget } from './build';
import { ReadmeFile } from './files';
import { LintTarget, LintOptions } from './lint';
import { TestTarget, TestOptions } from './test';
import { TaskOptionsTarget } from './tasks';
import { TS_NODE_VERSION } from './constants';

// https://peps.python.org/pep-0508/
const DEP_NAME_VERSION_REGEX = /^([A-Za-z0-9][A-Za-z0-9._-]*[A-Za-z0-9])(.*)$/;

/**
 * Python project with basic configurations for linting, testing, building etc
 * @pjid python_basic
 */
export class PythonBasicProject extends Project {
  constructor(options: PythonBasicOptions) {
    const optionsWithDefaults = getPythonBasicOptionsWithDefaults(options);
    super({
      ...options,
      name: resolvePackageName(optionsWithDefaults.name, optionsWithDefaults.package),
    });

    // add deps to project
    for (const dep of options.deps ?? []) {
      if (DEP_NAME_VERSION_REGEX.test(dep)) {
        this.addDep(dep);
      } else {
        this.logger.info(`Skipping invalid dep: ${dep}`);
      }
    }
    for (const dep of options.devDeps ?? []) {
      if (DEP_NAME_VERSION_REGEX.test(dep)) {
        this.addDevDep(dep);
      } else {
        this.logger.info(`Skipping invalid devDep: ${dep}`);
      }
    }

    const venvPath = optionsWithDefaults.venvPath ?? '.venv';

    // cleanup default tasks
    cleanupTasks(this);

    // create .projenrc.py
    new Projenrc(this, { pythonExec: `${venvPath}/bin/python` });

    // create README.md
    new ReadmeFile(this, {
      projectName: optionsWithDefaults.name,
      description: optionsWithDefaults.package?.description,
    });

    // create sample
    if (optionsWithDefaults.sample) {
      new PythonBasicSample(this);
    }

    // create Makefiles and .nvmrc
    new BaseTooling(this, {
      nodeVersion: NODE_VERSION,
      projenLibVersion: PROJEN_VERSION,
      tsNodeLibVersion: TS_NODE_VERSION,
      additionalMakefileContentsProjen: `prepare-venv:
	npx projen prepare-venv

`,
      additionalMakefileContentsTargets: {
        prepare: `
brew install python
brew install pyenv
make prepare-venv
make prepare-projen
`,
      },
    });

    // LINT
    new LintTarget(
      this,
      {
        venvPath,
        attachTasksTo: 'lint',
      },
      optionsWithDefaults.lint,
    );

    // TEST
    new TestTarget(
      this,
      {
        venvPath,
        attachTasksTo: 'test',
      },
      optionsWithDefaults.test,
    );

    // BUILD
    // This should be in the last position because other components might add dependencies
    // and they will be used to generate pyproject.toml, for example
    new BuildTarget(
      this,
      {
        venvPath,
        attachTasksTo: 'build',
      },
      {
        pip: optionsWithDefaults.pip,
        package: optionsWithDefaults.package,
      },
    );
  }

  /**
   * Add a runtime dependency in format [package-name][==version]
   * E.g: `addDep('package==1.0.0')`
   */
  public addDep(packageNameVersion: string): void {
    this.addDependency(packageNameVersion, DependencyType.RUNTIME);
  }

  /**
   * Add a development dependency in format [package-name][==version]
   */
  public addDevDep(packageNameVersion: string): void {
    this.addDependency(packageNameVersion, DependencyType.DEVENV);
  }

  private addDependency(packageNameVersion: string, type: DependencyType): void {
    // extract package name and version
    const match = DEP_NAME_VERSION_REGEX.exec(packageNameVersion);
    if (!match) {
      throw new Error(`Invalid package name/version: ${packageNameVersion}`);
    }
    const packageName = match[1];
    const versionSpec = match[2];
    this.deps.addDependency(`${packageName}${versionSpec ? '@' : ''}${versionSpec}`, type);
  }
}

export interface PythonBasicOptions extends ProjectOptions, Build0Options, TaskOptionsTarget {
  /**
   * Package dependencies in format `['package==1.0.0', 'package2==2.0.0']`
   */
  readonly deps?: string[];
  /**
   * Development dependencies in format `['package==1.0.0', 'package2==2.0.0']`
   */
  readonly devDeps?: string[];
  /**
   * Create sample code and test (if dir doesn't exist yet)
   * @default true
   */
  readonly sample?: boolean;
  /**
   * Linting configurations such as rules selected etc
   */
  readonly lint?: LintOptions;
  /**
   * Test configurations
   */
  readonly test?: TestOptions;
}

const getPythonBasicOptionsWithDefaults = (options: PythonBasicOptions): PythonBasicOptions => {
  return {
    sample: true,
    ...options,
  };
};

const cleanupTasks = (project: Project): void => {
  // cleanup default tasks
  project.tasks.removeTask('build');
  project.tasks.removeTask('pre-compile');
  project.tasks.removeTask('post-compile');
  project.tasks.removeTask('compile');
  project.tasks.removeTask('test');
  project.tasks.removeTask('package');
};
