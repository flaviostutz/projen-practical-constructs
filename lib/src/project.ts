/* eslint-disable no-restricted-syntax */
/* eslint-disable no-new */
import { DependencyType, License, LicenseOptions, Project, ProjectOptions } from 'projen';
import { Projenrc } from 'projen/lib/python';

import { PyProjectTomlFile, PyProjectTomlOptions, resolvePackageName } from './files/pyproject';
import { PythonVersionFile } from './files/python-version';
import { addDefaultGitIgnore } from './utils/addDefaultGitIgnore';
import { PythonBasicSample } from './sample';

export class PythonBasicProject extends Project {
  constructor(options: PythonBasicOptions) {
    const optionsWithDefaults = getPythonBasicOptionsWithDefaults(options);
    super({
      ...options,
      name: resolvePackageName(optionsWithDefaults.name, optionsWithDefaults.package),
    });

    // add deps to project
    for (const dep of options.deps ?? []) {
      this.addDep(dep);
    }
    for (const dep of options.devDeps ?? []) {
      this.addDevDep(dep);
    }

    // create .projenrc.py
    new Projenrc(this, { pythonExec: `${optionsWithDefaults.venvPath}/bin/python` });

    // create LICENSE
    if (options.license) {
      new License(this, options.license);
    }

    // create sample
    if (optionsWithDefaults.sample) {
      new PythonBasicSample(this);
    }

    addDefaultGitIgnore(this);

    // ADD THIS AT LAST BECAUSE DEPS IN project MIGHT BE ADDED BY OTHER COMPONENTS
    // and PyProjectTomlFile will add them to pyproject.toml
    if (optionsWithDefaults.package) {
      // create pyproject.toml
      new PyProjectTomlFile(this, optionsWithDefaults.package);
      // create .python-version
      if (optionsWithDefaults.package.requiresPython) {
        const requiresPython = optionsWithDefaults.package.requiresPython.replace(/[^0-9.]/g, '');
        new PythonVersionFile(this, { pythonVersion: requiresPython });
      }
    }
  }

  public addDep(depVersion: string): void {
    this.deps.addDependency(depVersion, DependencyType.RUNTIME);
  }

  public addDevDep(depVersion: string): void {
    this.deps.addDependency(depVersion, DependencyType.DEVENV);
  }
}

export interface PythonBasicOptions extends ProjectOptions {
  /**
   * Package dependencies in format `['package==1.0.0', 'package2==2.0.0']`
   */
  readonly deps?: string[];
  /**
   * Development dependencies in format `['package==1.0.0', 'package2==2.0.0']`
   */
  readonly devDeps?: string[];
  /**
   * Python package options
   */
  readonly package?: PyProjectTomlOptions;
  /**
   * License options
   */
  readonly license?: LicenseOptions;
  /**
   * Create sample code and test (if dir doesn't exist yet)
   * @default true
   */
  readonly sample?: boolean;
  /**
   * Path to the python virtual environment directory
   * used in this project
   */
  readonly venvPath?: string;
}

const getPythonBasicOptionsWithDefaults = (options: PythonBasicOptions): PythonBasicOptions => {
  const packageWithDefaults = {
    requiresPython: '>=3.12',
    ...options.package,
  };
  return {
    venvPath: '.venv',
    ...options,
    package: packageWithDefaults,
    sample: options.sample ?? true,
  };
};
