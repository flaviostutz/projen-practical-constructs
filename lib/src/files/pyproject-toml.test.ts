/* eslint-disable no-new */
import { tmpdir } from 'os';
import { join } from 'path';
import { mkdtempSync } from 'fs';

import { Testing } from 'projen';

import { PythonBasicProject } from '../project';

import { PyProjectTomlFile } from './pyproject-toml';

describe('pyproject', () => {
  test('invalid package name should fail', () => {
    const outdir = mkdtempSync(join(tmpdir(), 'test-'));
    const project = new PythonBasicProject({ name: 'test2', outdir });
    new PyProjectTomlFile(project, { packageName: '&^ABC' });
    expect(() => Testing.synth(project)).toThrow(/Invalid package name/);
    project.synth();
  });
  test('invalid version should fail', () => {
    const outdir = mkdtempSync(join(tmpdir(), 'test-'));
    const project = new PythonBasicProject({ name: 'test3', outdir });
    new PyProjectTomlFile(project, { version: 'AAAA' });
    project.synth();
    expect(() => Testing.synth(project)).toThrow(/Invalid version format/);
  });
});
