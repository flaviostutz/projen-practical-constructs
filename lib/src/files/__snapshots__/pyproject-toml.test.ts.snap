// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pyproject valid package name and version 1`] = `
"[build-system]
requires = [ "setuptools" ]
build-backend = "setuptools.build_meta"

[project]
name = "abc"
version = "1.0.0"
description = ""
readme = ""
requires-python = ">=3.12"
license = { file = "LICENSE" }
keywords = [ ]
dependencies = [ "helloworld==0.1dev" ]

[project.optional-dependencies]
dev = [ ]

[tool.setuptools.package-data]
abc = [ "py.typed" ]
"
`;
